File: day03/part2.py
_TESTCASE = """\
00100
11110
10110
10111
10101
01111
00111
11100
10000
11001
00010
01010
""".splitlines()


def compute(data):
    """
    >>> compute(_TESTCASE)
    230
    """

    def find_value(numbers, *, check_for_most_common):
        bit_cnt = len(numbers[0])

        for i in range(bit_cnt):
            one_bit_count = sum(1 for num in numbers if num[i] == "1")
            most_common = 1 if one_bit_count >= len(numbers) / 2 else 0

            if check_for_most_common:
                bit_check = str(most_common)
            else:
                bit_check = str(most_common ^ 1)

            numbers = [num for num in numbers if num[i] == bit_check]

            if len(numbers) == 1:
                break

        return int(numbers[0], 2)

    oxygen = find_value(data.copy(), check_for_most_common=True)
    co2 = find_value(data.copy(), check_for_most_common=False)

    return oxygen * co2


def main():
    import pathlib

    input_path = pathlib.Path(__file__).with_name("input.txt")

    with input_path.open() as f:
        print(compute(f.read().splitlines()))


if __name__ == "__main__":
    main()

File: day03/part2.cpp
#include <iostream>
#include <fstream>
#include <vector>
#include <algorithm>


int findValue(std::vector<std::string> numbers, bool checkForMostCommon) {
    auto bitCount = numbers[0].size();

    for (int i = 0; i < bitCount; ++i) {
        int oneBitCount = 0;
        for (auto num : numbers) {
            if (num[i] == '1') {
                oneBitCount++;
            }
        }

        int mostCommon = oneBitCount >= numbers.size() - oneBitCount;

        int numToCheck;
        if (checkForMostCommon) {
            numToCheck = mostCommon;
        } else {
            numToCheck = mostCommon ^ 1;
        }

        numbers.erase(
            std::remove_if(numbers.begin(), numbers.end(), [&](auto n) {
                return n[i] != std::to_string(numToCheck)[0];
            }),
            numbers.end()
        );

        if (numbers.size() == 1) {
            break;
        }
    }

    return stoi(numbers[0], nullptr, 2);
}


int compute(const std::vector<std::string>& numbers) {
    return findValue(numbers, true) * findValue(numbers, false);
}


int main() {
    std::string sourceFilePath = __FILE__;
    auto sourceDirPath = sourceFilePath.substr(0, sourceFilePath.rfind('\\'));
    auto inputPath = sourceDirPath + "/input.txt";

    std::ifstream file(inputPath);

    std::vector<std::string> numbers;

    if (!file) {
        std::cerr << "Couldn't open file " << inputPath << std::endl;
        return 1;
    }

    if (file.is_open()) {
        std::string line;
        while (std::getline(file, line)) {
            if (!line.empty()) {
                numbers.push_back(line);
            }
        }
        std::cout << compute(numbers) << std::endl;
        file.close();
    }

    return 0;
}
